src/corefx/System.Globalization.Native/icushim.cpp

```diff
@@ -0,0 +1,214 @@
+
+// Get filename of an ICU library with the requested version in the name
+// There are three possible cases of the version components values:
+// 1. Only majorVer is not equal to -1 => result is baseFileName.majorver
+// 2. Only majorVer and minorVer are not equal to -1 => result is baseFileName.majorver.minorVer
+// 3. All components are not equal to -1 => result is baseFileName.majorver.minorVer.subver
+void GetVersionedLibFileName(const char* baseFileName, int majorVer, int minorVer, int subVer, char* result)
+{
+    assert(majorVer != -1);
+
+    int nameLen = sprintf(result, "%s.%d", baseFileName, majorVer);
+
+    if (minorVer != -1)
+    {
+        nameLen += sprintf(result + nameLen, ".%d", minorVer);
+        if (subVer != -1)
+        {
+            sprintf(result + nameLen, ".%d", subVer);
+        }    
+    }
+}
+
+// Try to open the necessary ICU libraries
+bool OpenICULibraries(int majorVer, int minorVer, int subVer)
+{
+    char libicuucName[64];
+    char libicui18nName[64];
+
+    GetVersionedLibFileName("libicuuc.so", majorVer, minorVer, subVer, libicuucName);
+    GetVersionedLibFileName("libicui18n.so", majorVer, minorVer, subVer, libicui18nName);
+
+    libicuuc = dlopen(libicuucName, RTLD_LAZY);
+    if (libicuuc != nullptr)
+    {
+        libicui18n = dlopen(libicui18nName, RTLD_LAZY);
+        if (libicui18n == nullptr)
+        {
+            dlclose(libicuuc);
+            libicuuc = nullptr;
+        }
+    }
+
+    return libicuuc != nullptr;
+}
+
+// Select libraries using the version override specified by the CLR_ICU_VERSION_OVERRIDE
+// environment variable.
+// The format of the string in this variable is majorVer[.minorVer[.subVer]] (the brackets
+// indicate optional parts).
+bool FindLibUsingOverride(int* majorVer, int* minorVer, int* subVer)
+{
+    char* versionOverride = getenv("CLR_ICU_VERSION_OVERRIDE");
+    if (versionOverride != nullptr)
+    {
+        int first = -1;
+        int second = -1;
+        int third = -1;
+
+        int matches = sscanf(versionOverride, "%d.%d.%d", &first, &second, &third);
+        if (matches > 0)
+        {
+            if (OpenICULibraries(first, second, third))
+            {
+                *majorVer = first;
+                *minorVer = second;
+                *subVer = third;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+// Select the highest supported version of ICU present on the local machine
+// Search for library files with names including the major and minor version.
+bool FindLibWithMajorMinorVersion(int* majorVer, int* minorVer)
+{
+    for (int i = MaxICUVersion; i >= MinICUVersion; i--)
+    {
+        for (int j = MaxMinorICUVersion; j >= MinMinorICUVersion; j--)
+        {
+            if (OpenICULibraries(i, j, -1))
+            {
+                *majorVer = i;
+                *minorVer = j;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
```

Comment:
I have tried to swap the order of the loops to start with the version that tries to go with the three components of the version, which fails to find anything on my Ubuntu. The whole initialization took 45ms in that case on my machine. Without this flip of order, it takes 8ms.